### PR TITLE
Enable OTEL tracing to capture diagnostics activity info from OPCUA server

### PR DESCRIPTION
**Pull request by bhnaphade**
## Purpose
Enable OTEL support on Opc-plc to capture Tracing information from OPC UA Server.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: src/Configuration/CliOptions.cs [hunk 1]
```diff
@@ -71,7 +71,12 @@ public static (PlcSimulation PlcSimulationInstance, List<string> ExtraArgs) Init
                         throw new OptionException("The max OPC string length must be larger than 0.", "opcmaxstringlen");
                     }
                 }
-            },
+            },
+
+            // OTLP configuration
+            { "otlpe|otlpendpoint=", $"the endpoint Uri to which the otlp exporter is going to send traces .\nDefault: {config.OtlpEndpointUri}", (s) => config.OtlpEndpointUri = s },
+            { "otlpei|otlpexportinterval=", $"the interval in seconds to export traces in milliseconds.\nDefault: {config.OtlpExportInterval}", (int i) => config.OtlpExportInterval = i },
```
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:78]: >in seconds to export traces in milliseconds [](http://example.com/codeflow?start=59&length=43)

in seconds in milliseconds?

File: src/Configuration/CliOptions.cs [hunk 2]
```diff
@@ -71,7 +71,12 @@ public static (PlcSimulation PlcSimulationInstance, List<string> ExtraArgs) Init
                         throw new OptionException("The max OPC string length must be larger than 0.", "opcmaxstringlen");
                     }
                 }
-            },
+            },
+
+            // OTLP configuration
+            { "otlpe|otlpendpoint=", $"the endpoint Uri to which the otlp exporter is going to send traces .\nDefault: {config.OtlpEndpointUri}", (s) => config.OtlpEndpointUri = s },
```
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:77]: >  [](http://example.com/codeflow?start=106&length=1)

Nit: remove blank

Comment (luiscantero) [inline src/Configuration/CliOptions.cs:77]: >Uri [](http://example.com/codeflow?start=52&length=3)

Nit: URI

File: src/Configuration/Configuration.cs [hunk 3]
```diff
@@ -35,6 +36,16 @@ public class OpcPlcConfiguration
...
 
+    /// <summary>
+    /// Gets or sets OTLP reporting endpoint URI.
+    /// </summary>
+    public string OtlpEndpointUri { get; set; } = "http://localhost:4317";
+
+    /// <summary>
+    /// Gets or sets a the export interval in milliseconds.
+    /// </summary>
+    public int OtlpExportInterval { get; set; } = (int)TimeSpan.FromSeconds(60).TotalMilliseconds;
```
Comment (luiscantero) [inline src/Configuration/Configuration.cs:47]: >int [](http://example.com/codeflow?start=11&length=3)

Could we make this a TimeSpan and when we use it convert it to ms?

File: src/Configuration/CliOptions.cs (hunk 2 above)
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:77]: >is going to send traces [](http://example.com/codeflow?start=83&length=23)

I don't think the exporter sends traces to an endpoint, it rather makes them available there, no?

File: src/OpcPlcServer.cs [hunk 4]
```diff
@@ -164,24 +169,39 @@
...
-            host.Start();
-
-            if (Config.ShowPublisherConfigJsonIp)
-            {
-                Logger.LogInformation("Web server started: {pnJsonUri}", $"http://{GetIpAddress()}:{Config.WebServerPort}/{Config.PnJson}");
-            }
-            else if (Config.ShowPublisherConfigJsonPh)
-            {
-                Logger.LogInformation("Web server started: {pnJsonUri}", $"http://{Config.OpcUa.Hostname}:{Config.WebServerPort}/{Config.PnJson}");
-            }
```
Comment (luiscantero) [inline src/OpcPlcServer.cs:180]: We still need to show this information. Did you check if the pn.json can still be retrieved?

File: src/Configuration/CliOptions.cs (hunk 2 above)
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:77]: >otlpendpoint [](http://example.com/codeflow?start=21&length=12)

Do we need to specify the endpoint? shouldn't the port be enough?

We also have an option to specify the port for the pn.json, we need to figure that out.

Comment (bhnaphade) [reply src/Configuration/CliOptions.cs]: updated

File: src/OpcPlcServer.cs (hunk 4 above)
Comment (bhnaphade) [reply src/OpcPlcServer.cs]: restored web application hosting pn.json file

File: src/Configuration/CliOptions.cs (hunk 2 above)
Comment (bhnaphade) [reply src/Configuration/CliOptions.cs]: endpoint is needed for the otel collector.

**Comment by bhnaphade** [comment]
This pull request is awaiting opc ua stack update nuget to be available.

File: src/Startup.cs [hunk 5]
```diff
@@ -47,5 +55,21 @@
...
+    // This method configures OpenTelemetry tracing.
+    public void ConfigureOpenTelemetryTracing()
+    {
+        var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddAspNetCoreInstrumentation()
+            .AddSource(EndpointBase.ActivitySourceName)
+            .SetResourceBuilder(ResourceBuilder.CreateDefault()
+                .AddService(Program.OpcPlcServer.Config.ProgramName))
+            .AddOtlpExporter(opt => {
+                opt.Endpoint = new Uri("http://otel-collector.opcuabroker-monitoring.svc.cluster.local:4317");
```
Comment (luiscantero) [inline src/Startup.cs:69]: >Uri [](http://example.com/codeflow?start=35&length=3)

Shouldn't this uri come from the config?

File: src/OpcPlcServer.cs [hunk 6]
```diff
@@ -91,15 +95,24 @@
         Logger.LogInformation("Log file: {logFileName}", Path.GetFullPath(Config.LogFileName));
         Logger.LogInformation("Log level: {logLevel}", Config.LogLevelCli);
 
-        // Show version.
+        // Show OPC PLC version.
```
Comment (luiscantero) [inline src/OpcPlcServer.cs:98]: >Show [](http://example.com/codeflow?start=11&length=4)

Pls check why this changes are showing up, main already contains them. Maybe a fwd integration mistake?

File: src/Startup.cs (hunk 5 above)
Comment (bhnaphade) [reply src/Startup.cs]: yes ,removed debug code

File: src/OpcPlcServer.cs (hunk 6 above)
Comment (bhnaphade) [reply src/OpcPlcServer.cs]: fixed merge with main

File: src/Configuration/CliOptions.cs [hunk 7]
```diff
@@ -71,7 +71,23 @@ public static (PlcSimulation PlcSimulationInstance, List<string> ExtraArgs) Init
                         throw new OptionException("The max OPC string length must be larger than 0.", "opcmaxstringlen");
                     }
                 }
-            },
+            },
+
+            // OTLP Exporter Configuration
+            { "otlpen|otlpenabled", $"enables the otlp exporter.\nDefault: {config.OtlpEnabled}", (s) => config.OtlpEnabled = s != null },
+            { "otlpep|otlpendpoint=", $"the endpoint URI to which the otlp exporter is going to send traces.\nDefault: {config.OtlpEndpointUri}", (s) => config.OtlpEndpointUri = s },
+            { "otlpei|otlpexportinterval=", $"the interval to export traces in seconds.\nDefault: {config.OtlpExportInterval}", (int i) => {
```
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:79]: >int [](http://example.com/codeflow?start=129&length=3)

You can set this to uint to be consistent with other options below. Then we don't need the value check below

File: src/Configuration/CliOptions.cs [hunk 8]
```diff
@@ -71,7 +71,23 @@ public static (PlcSimulation PlcSimulationInstance, List<string> ExtraArgs) Init
                         throw new OptionException("The max OPC string length must be larger than 0.", "opcmaxstringlen");
                     }
                 }
-            },
+            },
+
+            // OTLP Exporter Configuration
+            { "otlpen|otlpenabled", $"enables the otlp exporter.\nDefault: {config.OtlpEnabled}", (s) => config.OtlpEnabled = s != null },
```
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:77]: >otlpen [](http://example.com/codeflow?start=15&length=6)

Do we need this to be enabled/disabled or could we use setting the endpoint as a way to enable it?

Comment (luiscantero) [reply src/Configuration/CliOptions.cs:77]: (depends on the requirements)

File: src/Startup.cs [hunk 9]
```diff
@@ -47,5 +55,21 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+    {
+        var tracerProvider = Sdk.CreateTracerProviderBuilder()
+            .AddAspNetCoreInstrumentation()
+            .AddSource(EndpointBase.ActivitySourceName)
+            .SetResourceBuilder(ResourceBuilder.CreateDefault()
+                .AddService(Program.OpcPlcServer.Config.ProgramName))
+            .AddOtlpExporter(opt => {
+                opt.Endpoint = new Uri(Program.OpcPlcServer.Config.OtlpEndpointUri);
+                opt.Protocol = OtlpExportProtocol.Grpc;
+                opt.BatchExportProcessorOptions.ExporterTimeoutMilliseconds = Program.OpcPlcServer.Config.OtlpExportInterval.Milliseconds;
```
Comment (luiscantero) [inline src/Startup.cs:71]: >Milliseconds [](http://example.com/codeflow?start=125&length=12)

This has to be TotalMilliseconds

File: src/Configuration/CliOptions.cs (hunk 8 above)
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:77]: >otlp [](http://example.com/codeflow?start=50&length=4)

Nit: let's spell it as OTLP, also below

File: src/opc-plc.csproj [hunk 10]
```diff
@@ -25,6 +25,15 @@
     <FrameworkReference Include="Microsoft.AspNetCore.App" />
     <PackageReference Include="Microsoft.Extensions.Logging" Version="8.0.0" />
     <PackageReference Include="Mono.Options" Version="6.12.0.148" />
+    <PackageReference Include="OpenTelemetry" Version="1.7.0" />
+    <PackageReference Include="OpenTelemetry.Api" Version="1.7.0" />
+    <PackageReference Include="OpenTelemetry.Api.ProviderBuilderExtensions" Version="1.7.0" />
+    <PackageReference Include="OpenTelemetry.Exporter.Console" Version="1.7.0" />
```
Comment (luiscantero) [inline src/opc-plc.csproj:31]: >PackageReference [](http://example.com/codeflow?start=5&length=16)

Pls update the version.json file to 2.12.4

File: src/Configuration/CliOptions.cs [hunk 11]
```diff
@@ -71,7 +71,23 @@ public static (PlcSimulation PlcSimulationInstance, List<string> ExtraArgs) Init
...
                     }
                 }
-            },
+            },
+
+            // OTLP Exporter Configuration
+            { "otlpen|otlpenabled", $"enables the otlp exporter.\nDefault: {config.OtlpEnabled}", (s) => config.OtlpEnabled = s != null },
+            { "otlpep|otlpendpoint=", $"the endpoint URI to which the otlp exporter is going to send traces.\nDefault: {config.OtlpEndpointUri}", (s) => config.OtlpEndpointUri = s },
+            { "otlpei|otlpexportinterval=", $"the interval to export traces in seconds.\nDefault: {config.OtlpExportInterval}", (int i) => {
+                    if (i > 0)
```
Comment (felixse) [inline src/Configuration/CliOptions.cs:80]: I suggest to add another option to specify the export protocol (grpc or protobuf)

File: src/Configuration/CliOptions.cs [hunk 12]
```diff
@@ -71,7 +71,13 @@ public static (PlcSimulation PlcSimulationInstance, List<string> ExtraArgs) Init
                         throw new OptionException("The max OPC string length must be larger than 0.", "opcmaxstringlen");
                     }
                 }
-            },
+            },
+
+            // OTLP Exporter Configuration
+            { "otlpen|otlpenabled", $"enables the otlp exporter.\nDefault: {config.OtlpEnabled}", (s) => config.OtlpEnabled = s != null },
```
Comment (luiscantero) [inline src/Configuration/CliOptions.cs:77]:
```suggestion
            { "otlpen|otlpenabled", $"enables the OTLP exporter.\nDefault: {config.OtlpEnabled}", (s) => config.OtlpEnabled = s != null },
```

File: src/Configuration/CliOptions.cs (hunk 11 above)
Comment (luiscantero) [reply src/Configuration/CliOptions.cs]: Done, txs

File: src/Configuration/CliOptions.cs [hunk 13]
```diff
@@ -71,7 +71,23 @@ public static (PlcSimulation PlcSimulationInstance, List<string> ExtraArgs) Init
...
+            { "otlpep|otlpendpoint=", $"the endpoint URI to which the otlp exporter is going to send traces.\nDefault: {config.OtlpEndpointUri}", (s) => config.OtlpEndpointUri = s },
```
[the remark this answers is not in the document]
Comment (luiscantero) [reply src/Configuration/CliOptions.cs]: Done

**Review by felixse — APPROVED** [review]

**Review by luiscantero — APPROVED** [review]
